### DISCUSSION TITLE
node_modules/.pnpm/node_modules

### DISCUSSION THREAD

**Discussion by samchungy** [pnpm/pnpm#4558] [Q&A] [answered]
👋 Hello, I'm trying to fix the pnpm integration for serverless-esbuild.

What's the purpose of `node_modules/.pnpm/node_modules` folder? Does it actually get used? I can see that if I have multiple dependencies of the same name but different versions only one ends up there.

How do I determine which version of a dependency ends up in that folder?

**Comment by zkochan** [comment, marked as answer]
The [hoist-pattern](https://pnpm.io/npmrc#hoist-pattern) setting controls which packages are hoisted to that directory.

pnpm just walks the dependency graph and hoist the first version of a package that is found.

**Reply by samchungy** [reply]
Legend! thank you, that makes sense. Thanks for the quick reply 💛

**Reply by quanru** [reply]
This can lead to inconsistent versions of hoisted packages between full and partial installations. How to solve it?

**Reply by zkochan** [reply]
There is no way to solve it currently. If you need filtering, consider setting `hoist` to `false`. There are other way to fix issues with phantom deps. For instance, you may use [packageExtensions](https://pnpm.io/package_json#pnpmpackageextensions)
